sdk/docs/how-to/build-ensemblers/01-pyfunc-interface.md

```diff
@@ -0,0 +1,122 @@
+
+In order to build a pyfunc ensembler and register it with Turing API, you would need to implement the 
+`turing.ensembler.PyFunc` interface. This requires implementation two methods, `initialize` and `ensemble`.
+
+```python
+class MyEnsembler(turing.ensembler.PyFunc):
+    """
+    Simple ensembler, that returns predictions from the `model_odd`
+    if `customer_id` is odd, or predictions from `model_even` otherwise
+    """
+
+    def initialize(self, artifacts: dict):
+        pass
+
+    def ensemble(
+            self,
+            features: pandas.Series,
+            predictions: pandas.Series,
+            treatment_config: Optional[dict]) -> Any:
+        customer_id = features["customer_id"]
+        if (customer_id % 2) == 0:
+            return predictions['model_even']
+        else:
+            return predictions['model_odd']
+```
+
+`initialize` is essentially a method that gets called when an object of your implemented class gets instantiated. 
+
+`ensemble` on the other hand, works a little differently depending on whether you plan to implement the 
+`turing.ensembler.PyFunc` interface as a [batch ensembler](#batch-ensembler) or a 
+[real-time ensembler service](#real-time-ensembler-service).
+
+## Batch Ensembler
+As a batch ensembler, `ensemble` will be called when performing ensembling jobs.
+
+The source and predictions columns specified in your ensembling job will be passed **row-by-row** to `ensemble` as the 
+arguments `features` and `predictions` respectively. These two arguments are `pandas.Series` objects which you can 
+manipulate freely with in your implementation of the `ensemble` method. 
+
+Note that for the moment, the argument `treatment_config` is not supported in batch ensembling; i.e. it will be 
+passed as `None` in each call of `ensemble`.
+
+The results of `ensemble` will then be stored in the sink that you have specified as part of the ensembling job.
+
+The following example shows how the `ensemble` method may look like when being implemented for batch ensembling: 
+```python
+def ensemble(
+        self,
+        features: pandas.Series,
+        predictions: pandas.Series,
+        treatment_config: Optional[dict]) -> Any:
+    customer_id = features["customer_id"]
+    if (customer_id % 2) == 0:
+        return predictions['model_even']
+    else:
+        return predictions['model_odd']
+```
+
+## Real-Time Ensembler Service
+As a real-time ensembler service, `ensemble` will be called each time an individual request gets sent to the 
+ensembler service (which gets automatically built and run as part of a Turing router deployment).
+
+Each time a Turing Router sends a request to a 
+[pyfunc ensembler service](../../../../docs/how-to/create-a-router/configure-ensembler.md#pyfunc-ensembler), 
+`ensemble` will be called, with the request payload being passed as a `dict` object for the `features` argument, and 
+the route responses as a `list` of `dict` for the `predictions` argument. 
```

Comment:
Actually the route responses will get returned as a `List`. Each element corresponds to a route response, which itself is a `Dict` which has the schema:

```python
{
  "route": // ... Route name,
  "data": {
    // ... Successful route response
  },
  "is_default": // ... Boolean value
}
```

I suppose you were wondering if the underlying implementation of the pyfunc ensembler engine helps transform the raw route responses into something more intuitive, i.e. a `Dict` with key-value pairs where the key's a route name and the value's the route response. In this case the answer's no because I thought preserving the original request schema would be most easily understood by users (this would be how a user would expect a request if he/she was building a Docker ensembler), but that said, we can also implement the transformation in the engine if you think it's more intuitive for users.